README.md
# Ohjelmistotekniikka, harjoitustyö

Deep Learning Anti-Cheat (DLAC) on projekti joka analysoi pelaajia CSGO pelissä ja ennustaa käyttääkö pelaaja jotain avustusta jota ei saisi. 

## HUOM vaatimusmäärittelystä löytyy parempi selitys projektista.



## Asennus
- [Release viikko 5](https://github.com/LaihoE/ot-harjoitustyo/releases/tag/Viikko5)

```bash
poetry install
```


## Voit kokeilla sovellusta ajamalla
```bash
poetry run invoke start
```
Tämän pitäisi avata file explorer jolla pysyt navigoimaan src/csvs kansioon 
ja valita csv tiedosto. Parin sekunnin jälkeen pitäisi avautua seuraavanlainen ikkuna:
![](./dokumentaatio/kuvat/example_output.png)



Voimme myös käydä tarkistamassa pelaajan profiilin Steamin kautta:
https://steamcommunity.com/profiles/76561199056381450/

Kappas vaan, pelaajalla on pelikielto sopivasti pelin aikoihin...

## Testit
```
poetry run invoke test
```

## Testikattavuusraportti
```
poetry run invoke coverage-report
```

## Dokumentaatio
- [Vaatimusmäärittely](./dokumentaatio/vaatimusmaarittely.md)
- [Tuntikirjanpito](./dokumentaatio/tuntikirjanpito.md)
- [Changelog](./dokumentaatio/changelog.md)
- [Arkkitehtuuri](./dokumentaatio/arkkitehtuuri.md)


dokumentaatio/changelog.md
# Changelog

## Viikko 3
- Käyttäjä pytstyy saamaan ennustukset terminaaliin
- Testattu että ML-algoritmi latautuu odotetusti
  
## Viikko 4
- Tieto voidaan tallentaa sqlite tietokantaan, sekä saada se sieltä ulos
- Käyttäjälle aukeaa file explorer josta voi valita csv tiedoston

# Viikko 5
- Integraatio GUI:n ja tietokannan välille
- Ennustukset menevät suoraan tietokantaan ja haetaan GUI:n avulla ulos

# Viikko 6
- Käyttäjä voi valita haluaako käyttää terminal varianttia vai normaalia GUI varianttia.

src/tests/database_test.py
import unittest
from anti_cheat import Model
import numpy as np
import os
from db import Database
import sqlalchemy
import sqlite3
import pandas as pd


class TestDb(unittest.TestCase):
    def setUp(self):
        dbpath = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..', 'database', 'database.db'))
        db = Database(dbpath)
        if not os.path.isfile(dbpath):
            data_dict = {}
            data_dict["predictions"] = [0.46584928035736084]
            data_dict["player_names"] = ["Two pa cool"]
            data_dict["player_ids"] = [76561198112665670]
            data_dict["ticks"] = [1384]
            data_dict["file_names"] = ["match730_003416179073414595204_0172264905_181.dem"]
            df = pd.DataFrame.from_dict(data_dict)
            df.to_sql('testtable', con=db.engine, if_exists='append')

    def test_insert_successful(self):
        dbpath = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..', 'database', 'database.db'))
        db = Database(dbpath)

        data_dict = {}
        data_dict["predictions"] = [0.46584928035736084]
        data_dict["player_names"] = ["Two pa cool"]
        data_dict["player_ids"] = [76561198112665670]
        data_dict["ticks"] = [1384]
        data_dict["file_names"] = ["match730_003416179073414595204_0172264905_181.dem"]
        df = pd.DataFrame.from_dict(data_dict)
        df.to_sql('testtable', con=db.engine, if_exists='append')

    def test_fetch_successful(self):
        """
        Seems like the setUp doesnt commit the changes? something is up, have to redo the insert here for it to work always.
        Maybe df.to_sql aint too great, will revisit next week
        """
        # insert
        dbpath = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'database', 'database.db'))
        db = Database(dbpath)
        data_dict = {}
        data_dict["predictions"] = [0.46584928035736084]
        data_dict["player_names"] = ["Two pa cool"]
        data_dict["player_ids"] = [76561198112665670]
        data_dict["ticks"] = [1384]
        data_dict["file_names"] = ["match730_003416179073414595204_0172264905_181.dem"]
        df = pd.DataFrame.from_dict(data_dict)
        df.to_sql('testtable', con=db.engine, if_exists='append')

        # Actual test
        data = db.find_data(table='testtable')
        row = data.iloc[0].tolist()
        # Each row has 6 values
        self.assertEqual(len(row), 5)
        # Tick is an integer
        self.assertEqual(type(int(row[-1])), int)
        # Steamid is correct
        self.assertEqual(len(str(row[2])), len('76561198978973136'))
        self.assertEqual(str(row[2])[:3], '765')
        # Probablity (confidence) is between 0 and 1
        self.assertEqual(0 < float(row[0]) < 1, True)


src/tests/anti_cheat_test.py
import unittest
from anti_cheat import Model
import numpy as np
import os


class TestModel(unittest.TestCase):
    def setUp(self):
        pass

    def test_model_initialized_succesful(self):
        # init model
        dirname = os.path.dirname(__file__)
        model_path = os.path.join(dirname, '..', 'utils', 'ml_model.onnx')
        model = Model(model_path)
        # Create correct shape array of ones and predict (1 sample with 192 rows and 5 columns)
        testdata = np.float32(np.ones((1, 192, 5)))
        ort_inputs = {model.ort_session.get_inputs()[0].name: testdata}
        ort_outs = np.array(model.ort_session.run(None, ort_inputs))
        cheating_confidence = list(ort_outs[0][:, 1])
        # Make sure the model gives expected confidence. Round to 5 decimals due to margin of error
        self.assertEqual(round(np.float32(cheating_confidence[0]), 5), round(
            np.float32(0.03877572), 5))

    def test_predict_to_sql(self):
        dirname = os.path.dirname(__file__)
        model_path = os.path.join(dirname, '..', 'utils', 'ml_model.onnx')
        csv_path = os.path.join(dirname, '..', 'csvs', 'data.csv')
        model = Model(model_path)
        success = model.predict_to_sql(csv_path)
        self.assertEqual(success, True)


laskarit/viikko2/unicafe/src/tests/kassapaate_test.py
import unittest
from kassapaate import Kassapaate
from maksukortti import Maksukortti

class TestKassapaate(unittest.TestCase):
    def setUp(self):
        self.kassapaate = Kassapaate()
        self.maksukortti = Maksukortti(0)

    def test_kassapaate_luotu_oikein(self):
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)
        self.assertEqual(self.kassapaate.edulliset, 0)
        self.assertEqual(self.kassapaate.maukkaat, 0)



    def test_edullisesti_kateisella_maksu_riittava(self):
        takaisin_rahaa = self.kassapaate.syo_edullisesti_kateisella(260)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100240)
        self.assertEqual(takaisin_rahaa, 20)
        self.assertEqual(self.kassapaate.edulliset, 1)

    def test_edullisesti_kateisella_maksu_ei_riita(self):
        takaisin_rahaa = self.kassapaate.syo_edullisesti_kateisella(42)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)
        self.assertEqual(takaisin_rahaa, 42)
        self.assertEqual(self.kassapaate.edulliset, 0)
 


 
    def test_maukkaasti_kateisella_maksu_riittava(self):
        takaisin_rahaa = self.kassapaate.syo_maukkaasti_kateisella(500)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100400)
        self.assertEqual(takaisin_rahaa, 100)
        self.assertEqual(self.kassapaate.maukkaat, 1)

    def test_maukkaasti_kateisella_maksu_ei_riita(self):
        takaisin_rahaa = self.kassapaate.syo_maukkaasti_kateisella(42)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)
        self.assertEqual(takaisin_rahaa, 42)
        self.assertEqual(self.kassapaate.maukkaat, 0)
    


    
    def test_edullisesti_kortilla_maksu_riittava(self):
        self.maksukortti.lataa_rahaa(1000)
        onnistui = self.kassapaate.syo_edullisesti_kortilla(self.maksukortti)
        self.assertEqual(onnistui, True)
        self.assertEqual(self.maksukortti.saldo, 760)
        self.assertEqual(self.kassapaate.edulliset, 1)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)


    def test_edullisesti_kortilla_maksu_ei_riita(self):
        onnistui = self.kassapaate.syo_edullisesti_kortilla(self.maksukortti)
        self.assertEqual(onnistui, False)
        self.assertEqual(self.kassapaate.edulliset, 0)
        self.assertEqual(self.maksukortti.saldo, 0)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)

    
    def test_maukkaasti_kortilla_maksu_riittava(self):
        self.maksukortti.lataa_rahaa(1000)
        onnistui = self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti)
        self.assertEqual(onnistui, True)
        self.assertEqual(self.maksukortti.saldo, 600)
        self.assertEqual(self.kassapaate.maukkaat, 1)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)




    def test_maukkaasti_kortilla_maksu_ei_riita(self):
        onnistui = self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti)
        self.assertEqual(onnistui, False)
        self.assertEqual(self.maksukortti.saldo, False)
        self.assertEqual(self.kassapaate.maukkaat, 0)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)


    def test_kortille_rahan_lataus_rahaa_riittavasti(self):
        self.kassapaate.lataa_rahaa_kortille(self.maksukortti, 42)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100042)
        self.assertEqual(self.maksukortti.saldo, 42)
        

    def test_kortille_rahan_lataus_rahaa_ei_riittavasti(self):
        self.kassapaate.lataa_rahaa_kortille(self.maksukortti, -5)
        self.assertEqual(self.kassapaate.kassassa_rahaa, 100000)
        self.assertEqual(self.maksukortti.saldo, 0)












src/db.py
import sqlite3
import sqlalchemy
import pandas as pd


class Database():
    def __init__(self, db_path) -> None:
        """

        :param db_path: Path where db is
        """
        """
        Creates two connections, one for pythons native sqlite3 and one for
        sqlalchemy. Sqlalchemy is needed for df.to_sql()

        :param db_path: path to db file
        """
        self.conn = sqlite3.connect(db_path)
        self.engine = sqlalchemy.create_engine(
            f'sqlite:///{db_path}', echo=False)

    def find_data(self, table):
        """
        Query SQL database with pd.read_sql()

        :param table: Table where we query from
        :return: df
        """
        df = pd.read_sql(f'SELECT predictions, player_names, player_ids, file_names, ticks '
                         f'FROM {table} ORDER BY predictions DESC',self.engine)
        return df

    def insert_prediction(self, datadict, table):
        """
        Inserts predictions to sql using pandas df.to_sql()

        :param datadict: Dictionary containing all insertable data
        :param table: table name where we insert
        :return:
        """
        df = pd.DataFrame.from_dict(datadict)
        df.to_sql(table, con=self.engine, if_exists='append')


src/ui/gui.py
import time
import tkinter as tk
from tkinter import ttk
from tkinter import filedialog as fd
import os
from ..anti_cheat import Model
from tkinter import *
from pandastable import Table, TableModel
from ..db import Database


class UI:
    """
    This one might need some refreshing, just quickly implemented file explorer
    """
    def __init__(self):

        # init model
        dirname = os.path.dirname(__file__)
        model_path = os.path.join(dirname, '..', 'utils', 'ml_model.onnx')
        self.model = Model(model_path)
        # Init tkinter
        self.root = tk.Tk()
        self.root.title('DLAC')
        self.root.geometry('300x150')
        self.filename = None
        self.open_button = ttk.Button(
            self.root,
            text='Select csv file (game to analyse)',
            command=self.select_game
        )

    def select_game(self):
        filetypes = (
            ('CSV', '*.csv'),
        )
        dirname = os.path.dirname(__file__)
        csv_folder_path = os.path.join(dirname, 'csvs')
        self.filename = fd.askopenfilename(
            title='Select csv file (game to analyse)',
            initialdir=csv_folder_path,
            filetypes=filetypes,
        )
        self.root.destroy()

    def start(self):
        self.open_button.pack(expand=True)
        self.root.mainloop()
        if self.filename is not None:
            self.model.predict_to_sql(self.filename)
        else:
            raise FileExistsError("No file selected!")


    def generate_table(self):
        root = tk.Tk()
        f = Frame(root)
        f.pack(fill=BOTH, expand=1)

        dirname = os.path.dirname(__file__)
        path_to_db = os.path.join(dirname, '..', 'database', 'database.db')
        db = Database(path_to_db)
        df = db.find_data('mytable')

        self.table = Table(f, dataframe=df,
                                showtoolbar=True, showstatusbar=True)
        self.table.show()
        root.mainloop()

src/anti_cheat.py
import math
import os
import joblib
import pandas as pd
import numpy as np
import onnxruntime
from db import Database


class Model:
    def __init__(self, model_path):
        """
        Initializes Machine-learning model

        :param model_path:
        """
        self.ort_session = onnxruntime.InferenceSession(model_path)

    def _load_csv_data(self, file_path):
        """
        Loads csv file with pandas.

        :param file_path:
        :return: df
        """
        return pd.read_csv(file_path, header=None)

    def _transform_data(self, data):
        """
        Takes in raw csv-data and splits it into data that is used in the prediction
        and into data that is only used in the end. The function also transforms the
        prediction data.

        :param data:
        :return: ml_data, other data: ml data is a numpy array and other is dict
        """
        # Data comes in rows of 1x1728 (rolled out to a row) so we unroll the data
        data = data.to_numpy().reshape(-1, 192, 9)
        # All the data isn't used for the prediction but will be needed later so we slice it here.
        # Each value in the dictionary is a list
        # Each list in the dictionary is the same length
        other_data = {"file_names": data[:, 0, 0],
                      "player_names": data[:, 0, 1],
                      "player_ids": data[:, 0, 2],
                      "ticks": data[:, 0, 3]
                      }
        # Data that is input into the model.
        ml_data = np.float32(data[:, :, 4:])

        # We scale the data for better model performance (not important to understand),
        # if interested see more:
        # https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html
        dirname = os.path.dirname(__file__)
        path_to_scaler = os.path.join(dirname, 'utils', 'scaler.gz')
        scaler = joblib.load(path_to_scaler)

        # Reshape data into 2d and back to 3d after scaling.
        # Pylint seems to complain about this one (false-positive?)
        # Disabling pylint lines wasn't liked so i left it in
        ml_data = ml_data.reshape((-1, 5))
        ml_data = scaler.transform(ml_data)
        ml_data = ml_data.reshape((-1, 192, 5))

        return ml_data, other_data

    def _predict(self, file_path, batch_size):
        """
        Does the actual prediction. Returns predictions combined with "other" data. Throws away
        data used in prediction after it's done.
        Predictions are done in batches to fit into RAM/VRAM


        :param file_path: Path to csv file
        :param batch_size: How many samples per prediction. Trade-off between memory and speed
        :return: other_data: A dictionary containing predictions and data related to the predictions
        """

        print("Reading data...")
        csv_data = self._load_csv_data(file_path)
        ml_data, other_data = self._transform_data(csv_data)

        # The predictions list will be the same length as the rest of the data
        # allowing easy indexing of them all
        total_batches = math.ceil(ml_data.shape[0] / batch_size)
        predictions = []
        print("Starting predictions...")
        for batch in range(total_batches):
            # Slice current batch
            data_this_batch = ml_data[batch_size *
                                      batch: batch_size * batch + batch_size, :, :]
            # Prep the data for input into ml model
            ort_inputs = {self.ort_session.get_inputs()
                [0].name: data_this_batch}
            # Does the actual prediction
            ort_outs = np.array(self.ort_session.run(None, ort_inputs))
            # index the cheating conf
            cheating_confidence = list(ort_outs[0][:, 1])
            # Add batch to predictions list
            predictions.extend(cheating_confidence)
        predictions = np.array(predictions)
        print("Predictions done")
        # Add predictions to other data
        # Now we dont need the ml_data anymore
        other_data["predictions"] = predictions
        return other_data

    def predict_to_terminal(self, file_name, threshold=0.97, batch_size=1000):
        """
        Predicts and outputs the predictions to terminal.

        :param file_name:
        :param threshold:
        :param batch_size: How many samples per prediction. Trade-off between memory and speed
        :return: None
        """

        data_dict = self._predict(file_name, batch_size)
        # Each key in data_dict is a feature for example who was the shooter or how likely the
        # player was to cheat during that shot. Each value in the dict is a list with the length
        # equal to the total amount of shots in the game

        # Print a header
        print("Prediction\t", "Player_name\t",
              "player_id\t", "tick\t", "file_name\t")

        for shot in range(len(data_dict["predictions"])):
            # If the shot was highly likely to be cheating
            if data_dict["predictions"][shot] > threshold:
                print(data_dict["predictions"][shot], "\t",
                      data_dict["player_names"][shot], "\t",
                      data_dict["player_ids"][shot], "\t",
                      data_dict["ticks"][shot], "\t",
                      data_dict["file_names"][shot])

    def predict_to_sql(self, file_name, batch_size=1000):
        """
        Takes in a file name and creates the predictions and inserts them into SQL.

        :param file_name: a csv file path
        :param batch_size: How many samples per prediction. Trade-off between memory and speed
        :return: success: True if everything works ok
        """
        data_dict = self._predict(file_name, batch_size)

        dirname = os.path.dirname(__file__)
        path_to_db = os.path.join(dirname, 'database', 'database.db')
        database = Database(path_to_db)
        database.insert_prediction(data_dict, 'mytable')
        return True
